File: backend/schemas/dejem.py
from datetime import date as Date
from datetime import datetime, time as Time
from enum import Enum

from pydantic import BaseModel, Field


class DejemMonthStatusEnum(str, Enum):
    OPEN_INTEREST = "OPEN_INTEREST"
    DISTRIBUTED_PENDING = "DISTRIBUTED_PENDING"
    DISTRIBUTED = "DISTRIBUTED"
    OPEN_SHIFTS = "OPEN_SHIFTS"
    FINISHED = "FINISHED"


class DejemShiftStatusEnum(str, Enum):
    OPEN = "OPEN"
    CLOSED = "CLOSED"
    READY_FOR_MAP = "READY_FOR_MAP"
    INTEGRATED = "INTEGRATED"
    FINISHED = "FINISHED"


class DejemShiftTypeEnum(str, Enum):
    FT = "FT"
    ROCAM = "ROCAM"
    OUTROS = "OUTROS"


class ParticipationTypeEnum(str, Enum):
    NORMAL = "NORMAL"
    EXTRAORDINARY = "EXTRAORDINARY"
    SUBSTITUTION = "SUBSTITUTION"


class ParticipantStatusEnum(str, Enum):
    REGISTERED = "REGISTERED"
    CONFIRMED = "CONFIRMED"
    CANCELLED = "CANCELLED"


# --- DejemMonth ---


class DejemMonthCreate(BaseModel):
    year: int = Field(ge=2000, le=2100)
    month: int = Field(ge=1, le=12)
    total_available_slots: int = Field(ge=0)
    monthly_limit_per_officer: int = Field(ge=1)


class DejemMonthUpdate(BaseModel):
    total_available_slots: int | None = Field(default=None, ge=0)
    monthly_limit_per_officer: int | None = Field(default=None, ge=1)


class DejemMonthPublic(BaseModel):
    id: int
    year: int
    month: int
    total_available_slots: int
    monthly_limit_per_officer: int
    status: DejemMonthStatusEnum
    created_by_id: int
    created_at: datetime
    updated_at: datetime
    interested_count: int = 0

    model_config = {"from_attributes": True}


# --- DejemInterest ---


class DejemInterestUpsert(BaseModel):
    interested: bool
    desired_slots: int = Field(default=0, ge=0)


class DejemInterestPublic(BaseModel):
    id: int
    month_id: int
    user_id: int
    interested: bool
    desired_slots: int
    created_at: datetime

    model_config = {"from_attributes": True}


class DejemInterestAdminRow(BaseModel):
    id: int
    month_id: int
    user_id: int
    interested: bool
    desired_slots: int
    created_at: datetime
    patente: str
    nome_guerra: str
    full_name: str | None
    role: str
    organizational_unit: str

    model_config = {"from_attributes": True}


# --- Distribution / Allocation ---


class DejemDistributionPreview(BaseModel):
    month_id: int
    total_available_slots: int
    interested_count: int
    monthly_limit_per_officer: int
    base_quantity: int
    remaining_after_base: int


class DejemAllocationPublic(BaseModel):
    id: int
    month_id: int
    user_id: int
    allocated_slots: int
    used_slots: int
    remaining_slots: int
    created_at: datetime

    model_config = {"from_attributes": True}


class DejemAllocationAdminRow(BaseModel):
    id: int
    month_id: int
    user_id: int
    allocated_slots: int
    used_slots: int
    remaining_slots: int
    created_at: datetime
    desired_slots: int
    patente: str
    nome_guerra: str
    full_name: str | None
    role: str
    organizational_unit: str
    display_order: int


class DejemDistributeResponse(BaseModel):
    month: DejemMonthPublic
    preview: DejemDistributionPreview
    leftover_slots: int
    allocations: list[DejemAllocationAdminRow]


# --- Shifts ---


class DejemShiftCreate(BaseModel):
    month_id: int
    date: Date
    start_time: Time
    end_time: Time
    shift_type: DejemShiftTypeEnum = DejemShiftTypeEnum.FT
    capacity: int = Field(ge=0)
    status: DejemShiftStatusEnum = DejemShiftStatusEnum.OPEN
    vehicle_id: int | None = None


class DejemShiftUpdate(BaseModel):
    date: Date | None = None
    start_time: Time | None = None
    end_time: Time | None = None
    shift_type: DejemShiftTypeEnum | None = None
    capacity: int | None = Field(default=None, ge=0)
    status: DejemShiftStatusEnum | None = None
    vehicle_id: int | None = None


class DejemShiftPublic(BaseModel):
    id: int
    month_id: int
    date: Date
    start_time: Time
    end_time: Time
    shift_type: DejemShiftTypeEnum
    capacity: int
    filled_slots: int = 0
    available_slots: int = 0
    status: DejemShiftStatusEnum
    vehicle_id: int | None = None
    vehicle_prefixo: str | None = None
    created_by_id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DejemShiftCalendarDay(BaseModel):
    date: Date
    shift_count: int
    total_capacity: int
    total_filled: int
    has_open: bool
    has_closed: bool
    has_finished: bool


class DejemShiftCalendarResponse(BaseModel):
    year: int
    month: int
    month_id: int | None
    days: list[DejemShiftCalendarDay]


class DejemShiftDayDetail(BaseModel):
    date: Date
    month_id: int | None
    shifts: list[DejemShiftPublic]


class DejemShiftDashboard(BaseModel):
    month_id: int
    year: int
    month: int
    total_shifts: int
    open_shifts: int
    closed_shifts: int
    finished_shifts: int
    integrated_shifts: int = 0
    total_capacity: int
    total_filled: int
    total_available: int
    avg_remaining_slots: float = 0.0


# --- Enrollment (fase 4.5) ---


class DejemAdminAddParticipant(BaseModel):
    user_id: int
    participation_type: ParticipationTypeEnum = ParticipationTypeEnum.NORMAL


class DejemParticipantAdminRow(BaseModel):
    id: int
    shift_id: int
    user_id: int
    participation_type: ParticipationTypeEnum
    status: ParticipantStatusEnum
    consumes_balance: bool
    created_at: datetime
    enrolled_by_id: int | None = None
    patente: str
    nome_guerra: str
    full_name: str | None = None
    remaining_slots: int = 0


class DejemEnrollmentResult(BaseModel):
    participant_id: int
    shift_id: int
    user_id: int
    participation_type: ParticipationTypeEnum
    status: ParticipantStatusEnum
    consumes_balance: bool
    remaining_slots: int | None = None
    created_at: datetime


class DejemMyShiftCard(BaseModel):
    id: int
    month_id: int
    date: Date
    start_time: Time
    end_time: Time
    shift_type: DejemShiftTypeEnum
    capacity: int
    filled_slots: int
    available_slots: int
    status: DejemShiftStatusEnum
    i_am_enrolled: bool = False
    my_participation_type: ParticipationTypeEnum | None = None


class DejemMyDayDetail(BaseModel):
    date: Date
    month_id: int | None
    shifts: list[DejemMyShiftCard]


class DejemMapMember(BaseModel):
    user_id: int
    patente: str
    nome_guerra: str
    display_order: int = 0


class DejemMapBlock(BaseModel):
    shift_id: int
    title: str
    shift_type: DejemShiftTypeEnum
    start_time: Time
    end_time: Time
    status: DejemShiftStatusEnum
    vehicle_prefixo: str | None = None
    members: list[DejemMapMember]


class DejemShiftTemplateCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    shift_type: DejemShiftTypeEnum = DejemShiftTypeEnum.FT
    start_time: Time
    end_time: Time
    default_capacity: int = Field(ge=0)
    is_active: bool = True


class DejemShiftTemplateUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    shift_type: DejemShiftTypeEnum | None = None
    start_time: Time | None = None
    end_time: Time | None = None
    default_capacity: int | None = Field(default=None, ge=0)
    is_active: bool | None = None


class DejemShiftTemplatePublic(BaseModel):
    id: int
    name: str
    shift_type: DejemShiftTypeEnum
    start_time: Time
    end_time: Time
    default_capacity: int
    is_active: bool
    created_by_id: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# --- Geração automática (fase 4.4.1) ---


class DejemMonthGenerateRequest(BaseModel):
    year: int = Field(ge=2000, le=2100)
    month: int = Field(ge=1, le=12)
    weekdays: list[int] = Field(
        min_length=1,
        description="Dias da semana: 0=segunda … 6=domingo (igual a date.weekday()).",
    )
    template_ids: list[int] = Field(min_length=1)
    replace_existing: bool = False
    ignore_holidays: bool = False  # reservado para fase futura


class DejemMonthGenerateResult(BaseModel):
    year: int
    month: int
    month_id: int
    created: int
    ignored: int
    replaced: int
    elapsed_ms: int


class DejemMonthGeneratePreviewItem(BaseModel):
    date: Date
    start_time: Time
    end_time: Time
    shift_type: DejemShiftTypeEnum
    capacity: int
    template_id: int
    template_name: str
    action: str  # CREATE | IGNORE | REPLACE
    status_label: str
    existing_shift_id: int | None = None


class DejemMonthGeneratePreview(BaseModel):
    year: int
    month: int
    month_id: int
    days_in_month: int
    selected_days_count: int
    weekdays: list[int]
    weekday_labels: list[str]
    template_names: list[str]
    replace_existing: bool
    planned_shifts: int
    planned_capacity: int
    create_count: int
    ignore_count: int
    replace_count: int
    create_capacity: int
    replace_capacity: int
    existing_conflicts: int
    items: list[DejemMonthGeneratePreviewItem]
    elapsed_ms: int

File: frontend/src/types/dejem.ts
export type DejemMonthStatus =
  | "OPEN_INTEREST"
  | "DISTRIBUTED_PENDING"
  | "DISTRIBUTED"
  | "OPEN_SHIFTS"
  | "FINISHED";

export type DejemShiftStatus =
  | "OPEN"
  | "CLOSED"
  | "READY_FOR_MAP"
  | "INTEGRATED"
  | "FINISHED";

export type DejemShiftType = "FT" | "ROCAM" | "OUTROS";

export type ParticipationType = "NORMAL" | "EXTRAORDINARY" | "SUBSTITUTION";

export type ParticipantStatus = "REGISTERED" | "CONFIRMED" | "CANCELLED";

export const DEJEM_MONTH_NAMES = [
  "",
  "Janeiro",
  "Fevereiro",
  "Março",
  "Abril",
  "Maio",
  "Junho",
  "Julho",
  "Agosto",
  "Setembro",
  "Outubro",
  "Novembro",
  "Dezembro",
] as const;

export function dejemMonthLabel(year: number, month: number): string {
  return `${DEJEM_MONTH_NAMES[month] ?? month}/${year}`.toUpperCase();
}

export const DEJEM_MONTH_STATUS_LABELS: Record<DejemMonthStatus, string> = {
  OPEN_INTEREST: "Manifestação aberta",
  DISTRIBUTED_PENDING: "Aguardando distribuição",
  DISTRIBUTED: "Distribuído",
  OPEN_SHIFTS: "Escalas abertas",
  FINISHED: "Encerrado",
};

export const DEJEM_SHIFT_STATUS_LABELS: Record<DejemShiftStatus, string> = {
  OPEN: "Aberta",
  CLOSED: "Fechada",
  READY_FOR_MAP: "Pronta p/ mapa",
  INTEGRATED: "No Mapa Força",
  FINISHED: "Finalizada",
};

export const DEJEM_SHIFT_TYPE_LABELS: Record<DejemShiftType, string> = {
  FT: "FT",
  ROCAM: "ROCAM",
  OUTROS: "Outros",
};

export const PARTICIPATION_TYPE_LABELS: Record<ParticipationType, string> = {
  NORMAL: "Normal",
  EXTRAORDINARY: "Extraordinária",
  SUBSTITUTION: "Substituição",
};

export const PARTICIPANT_STATUS_LABELS: Record<ParticipantStatus, string> = {
  REGISTERED: "Inscrito",
  CONFIRMED: "Confirmado",
  CANCELLED: "Cancelado",
};

export interface DejemMonthPublic {
  id: number;
  year: number;
  month: number;
  total_available_slots: number;
  monthly_limit_per_officer: number;
  status: DejemMonthStatus;
  created_by_id: number;
  created_at: string;
  updated_at: string;
  interested_count: number;
}

export interface DejemMonthCreatePayload {
  year: number;
  month: number;
  total_available_slots: number;
  monthly_limit_per_officer: number;
}

export interface DejemMonthUpdatePayload {
  total_available_slots?: number;
  monthly_limit_per_officer?: number;
}

export interface DejemInterestUpsertPayload {
  interested: boolean;
  desired_slots: number;
}

export interface DejemInterestPublic {
  id: number;
  month_id: number;
  user_id: number;
  interested: boolean;
  desired_slots: number;
  created_at: string;
}

export interface DejemInterestAdminRow {
  id: number;
  month_id: number;
  user_id: number;
  interested: boolean;
  desired_slots: number;
  created_at: string;
  patente: string;
  nome_guerra: string;
  full_name: string | null;
  role: string;
  organizational_unit: string;
}

export interface DejemDistributionPreview {
  month_id: number;
  total_available_slots: number;
  interested_count: number;
  monthly_limit_per_officer: number;
  base_quantity: number;
  remaining_after_base: number;
}

export interface DejemAllocationPublic {
  id: number;
  month_id: number;
  user_id: number;
  allocated_slots: number;
  used_slots: number;
  remaining_slots: number;
  created_at: string;
}

export interface DejemAllocationAdminRow {
  id: number;
  month_id: number;
  user_id: number;
  allocated_slots: number;
  used_slots: number;
  remaining_slots: number;
  created_at: string;
  desired_slots: number;
  patente: string;
  nome_guerra: string;
  full_name: string | null;
  role: string;
  organizational_unit: string;
  display_order: number;
}

export interface DejemDistributeResponse {
  month: DejemMonthPublic;
  preview: DejemDistributionPreview;
  leftover_slots: number;
  allocations: DejemAllocationAdminRow[];
}

export interface DejemShiftPublic {
  id: number;
  month_id: number;
  date: string;
  start_time: string;
  end_time: string;
  shift_type: DejemShiftType;
  capacity: number;
  filled_slots: number;
  available_slots: number;
  status: DejemShiftStatus;
  vehicle_id: number | null;
  vehicle_prefixo: string | null;
  created_by_id: number;
  created_at: string;
  updated_at: string;
}

export interface DejemShiftCreatePayload {
  month_id: number;
  date: string;
  start_time: string;
  end_time: string;
  shift_type: DejemShiftType;
  capacity: number;
  status?: DejemShiftStatus;
  vehicle_id?: number | null;
}

export interface DejemShiftUpdatePayload {
  date?: string;
  start_time?: string;
  end_time?: string;
  shift_type?: DejemShiftType;
  capacity?: number;
  status?: DejemShiftStatus;
  vehicle_id?: number | null;
}

export interface DejemShiftCalendarDay {
  date: string;
  shift_count: number;
  total_capacity: number;
  total_filled: number;
  has_open: boolean;
  has_closed: boolean;
  has_finished: boolean;
}

export interface DejemShiftCalendarResponse {
  year: number;
  month: number;
  month_id: number | null;
  days: DejemShiftCalendarDay[];
}

export interface DejemShiftDayDetail {
  date: string;
  month_id: number | null;
  shifts: DejemShiftPublic[];
}

export interface DejemShiftDashboard {
  month_id: number;
  year: number;
  month: number;
  total_shifts: number;
  open_shifts: number;
  closed_shifts: number;
  finished_shifts: number;
  total_capacity: number;
  total_filled: number;
  total_available: number;
  avg_remaining_slots?: number;
}

export interface DejemMyShiftCard {
  id: number;
  month_id: number;
  date: string;
  start_time: string;
  end_time: string;
  shift_type: DejemShiftType;
  capacity: number;
  filled_slots: number;
  available_slots: number;
  status: DejemShiftStatus;
  i_am_enrolled: boolean;
  my_participation_type: ParticipationType | null;
}

export interface DejemMyDayDetail {
  date: string;
  month_id: number | null;
  shifts: DejemMyShiftCard[];
}

export interface DejemParticipantAdminRow {
  id: number;
  shift_id: number;
  user_id: number;
  participation_type: ParticipationType;
  status: ParticipantStatus;
  consumes_balance: boolean;
  created_at: string;
  enrolled_by_id: number | null;
  patente: string;
  nome_guerra: string;
  full_name: string | null;
  remaining_slots: number;
}

export interface DejemAdminAddParticipantPayload {
  user_id: number;
  participation_type: ParticipationType;
}

export interface DejemEnrollmentResult {
  participant_id: number;
  shift_id: number;
  user_id: number;
  participation_type: ParticipationType;
  status: ParticipantStatus;
  consumes_balance: boolean;
  remaining_slots: number | null;
  created_at: string;
}

export interface DejemShiftTemplatePublic {
  id: number;
  name: string;
  shift_type: DejemShiftType;
  start_time: string;
  end_time: string;
  default_capacity: number;
  is_active: boolean;
  created_by_id: number;
  created_at: string;
  updated_at: string;
}

export interface DejemShiftTemplateCreatePayload {
  name: string;
  shift_type: DejemShiftType;
  start_time: string;
  end_time: string;
  default_capacity: number;
  is_active?: boolean;
}

export interface DejemShiftTemplateUpdatePayload {
  name?: string;
  shift_type?: DejemShiftType;
  start_time?: string;
  end_time?: string;
  default_capacity?: number;
  is_active?: boolean;
}

export interface DejemMonthGeneratePayload {
  year: number;
  month: number;
  weekdays: number[];
  template_ids: number[];
  replace_existing?: boolean;
  ignore_holidays?: boolean;
}

export interface DejemMonthGenerateResult {
  year: number;
  month: number;
  month_id: number;
  created: number;
  ignored: number;
  replaced: number;
  elapsed_ms: number;
}

export type DejemGeneratePreviewAction = "CREATE" | "IGNORE" | "REPLACE";

export interface DejemMonthGeneratePreviewItem {
  date: string;
  start_time: string;
  end_time: string;
  shift_type: DejemShiftType;
  capacity: number;
  template_id: number;
  template_name: string;
  action: DejemGeneratePreviewAction;
  status_label: string;
  existing_shift_id: number | null;
}

export interface DejemMonthGeneratePreview {
  year: number;
  month: number;
  month_id: number;
  days_in_month: number;
  selected_days_count: number;
  weekdays: number[];
  weekday_labels: string[];
  template_names: string[];
  replace_existing: boolean;
  planned_shifts: number;
  planned_capacity: number;
  create_count: number;
  ignore_count: number;
  replace_count: number;
  create_capacity: number;
  replace_capacity: number;
  existing_conflicts: number;
  items: DejemMonthGeneratePreviewItem[];
  elapsed_ms: number;
}

/** Normaliza horário da API (HH:MM:SS ou HH:MM) para input type=time. */
export function dejemTimeInputValue(value: string): string {
  return value.slice(0, 5);
}

export function formatDejemTime(value: string): string {
  return value.slice(0, 5);
}
